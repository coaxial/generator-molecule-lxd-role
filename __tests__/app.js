'use strict';
const { clone } = require('ramda');

const { readFileSync } = require('fs');
const assert = require('yeoman-assert');
const helpers = require('yeoman-test');
const path = require('path');

describe('generator-molecule-lxd-role:app', () => {
  const defaultResponses = {
    roleName: 'Test role',
    repoName: 'ansible-role-test-role',
    roleDesc: 'This is a test description for the role.',
    hasDeps: true,
    roleDeps:
      '- src: https://bithub.com/username/role.git\n- name: username.myrole\n  src: https://bithub.com/username/myrole.git',
    hasReqs: true,
    roleReqs: 'This role makes some assumptions.',
    license: 'MIT',
    authorName: 'Test author',
    authorOrganization: 'Test organization',
    authorWebsite: 'https://example.org',
    minAnsibleVer: '2.4',
    targetDistributions: ['UBUNTU'],
    galaxyTags: ['test', 'system', 'linux'],
    useTravis: true,
    travisUsername: 'test-travis-username',
    targetVersions: [
      {
        family: 'debian',
        distribution: 'ubuntu',
        codeName: 'trusty',
        versionNumber: '14.04',
        tags: ['lts', 'current'],
      },
      {
        family: 'debian',
        distribution: 'ubuntu',
        codeName: 'xenial',
        versionNumber: '16.04',
        tags: ['lts', 'current'],
      },
      {
        family: 'debian',
        distribution: 'ubuntu',
        codeName: 'bionic',
        versionNumber: '18.04',
        tags: ['lts', 'current'],
      },
    ],
  };

  describe('when all the prompts have answers', () => {
    const clonedResponses = clone(defaultResponses);

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      const filePath = 'README.md';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('.gitignore', () => {
      const filePath = '.gitignore';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('defaults/main.yml', () => {
      const filePath = 'defaults/main.yml';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('handlers/main.yml', () => {
      const filePath = 'handlers/main.yml';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('meta/main.yml', () => {
      const filePath = 'meta/main.yml';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('tasks/main.yml', () => {
      const filePath = 'tasks/main.yml';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('vars/main.yml', () => {
      const filePath = 'vars/main.yml';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('.travis/setup.sh', () => {
      const filePath = '.travis/setup.sh';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('.travis.yml', () => {
      const filePath = '.travis.yml';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('.yamllint', () => {
      const filePath = '.yamllint';

      it('exists', () => {
        assert.file(filePath);
      });

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('when there is no organization or website', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.authorOrganization = '';
    clonedResponses.authorWebsite = '';

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      it('formats the author information properly', () => {
        const actual = readFileSync('README.md', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('meta/main.yml', () => {
      it('formats the metadata properly', () => {
        const actual = readFileSync('meta/main.yml', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('when there is no organization', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.authorOrganization = '';

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      it('formats the author information properly', () => {
        const actual = readFileSync('README.md', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('meta/main.yml', () => {
      it('formats the metadata properly', () => {
        const actual = readFileSync('meta/main.yml', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('when there is no website', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.authorWebsite = '';

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      it('formats the author information properly', () => {
        const actual = readFileSync('README.md', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('meta/main.yml', () => {
      it('formats the metadata properly', () => {
        const actual = readFileSync('meta/main.yml', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('when there are no dependencies', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.hasDeps = false;
    clonedResponses.roleDeps = undefined;

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      it('skips the dependencies section', () => {
        const actual = readFileSync('README.md', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('meta/main.yml', () => {
      it('formats the metadata properly', () => {
        const actual = readFileSync('meta/main.yml', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('when there are no requirements', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.hasReqs = false;
    clonedResponses.roleReqs = undefined;

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      it('skips the requirents section', () => {
        const actual = readFileSync('README.md', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('meta/main.yml', () => {
      it('formats the metadata properly', () => {
        const actual = readFileSync('meta/main.yml', 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('when Travis disabled', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.useTravis = false;
    clonedResponses.travisUsername = undefined;

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('README.md', () => {
      const filePath = 'README.md';

      it('skips the badge', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });

    describe('.travis.yml', () => {
      const filePath = '.travis.yml';

      it('does not exist', () => {
        assert.noFile(filePath);
      });
    });
  });

  describe('when there are no galaxy tags', () => {
    const clonedResponses = clone(defaultResponses);
    clonedResponses.galaxyTags = [];

    beforeAll(() => {
      return helpers
        .run(path.join(__dirname, '../generators/app'))
        .withPrompts(clonedResponses);
    });

    describe('meta/main.yml', () => {
      const filePath = 'meta/main.yml';

      it('is correctly formatted', () => {
        const actual = readFileSync(filePath, 'utf8');

        expect(actual).toMatchSnapshot();
      });
    });
  });
});
