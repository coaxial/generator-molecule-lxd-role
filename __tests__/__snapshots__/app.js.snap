// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator-molecule-lxd-role:app when all the prompts have answers .gitignore is correctly formatted 1`] = `
"*.pyc
*.log
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers .travis.yml is correctly formatted 1`] = `
"---
jobs:
  include:
    - &test
      stage: test
      language: python
      python: \\"2.7\\"
      cache: pip
      dist: xenial
      sudo: required
      install:
        - .travis/setup.sh
      script:
        # the python venv needs to be in the path for molecule to run
        # https://unix.stackexchange.com/questions/83191/how-to-make-sudo-preserve-path
        - sudo env \\"PATH=$PATH\\" molecule test

notifications:
  webhooks: https://galaxy.ansible.com/api/v1/notifications/
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers .travis/setup.sh is correctly formatted 1`] = `
"#!/bin/sh
sudo apt-get update
# xenial ships LXD 2 LTS, backports ships 3 LTS
sudo apt-get upgrade -t xenial-backports -y lxd lxd-client

sudo lxd --version
# Wait for the socket to be ready
sudo lxd waitready
sudo lxd init --auto

pip install ansible
pip install molecule
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers README.md is correctly formatted 1`] = `
"Test role
=========
[![Build Status](https://travis-ci.org/coaxial/ansible-role-lxd.svg?branch=master)](https://travis-ci.org/coaxial/ansible-role-lxd)

This is a test description for the role.

Requirements
------------

This role makes some assumptions.

Role Variables
--------------

Name | Default | Possible values | Description
---|---|---|---
\`example_variable\` | \`true\` | \`true\` or \`false\` | This is an example to populate the table.

Dependencies
------------

- role
- username.myrole

Example Playbook
----------------

\`\`\`yaml
- hosts: all
  roles:
    - test-role
\`\`\`

License
-------

MIT

Author Information
------------------

Test author ([Test organization](https://example.org))"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers defaults/main.yml is correctly formatted 1`] = `
"---
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers handlers/main.yml is correctly formatted 1`] = `
"---
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers meta/main.yml is correctly formatted 1`] = `
"---
galaxy_info:
  author: Test author
  description: This is a test description for the role.
  company: Test organization
  license: MIT
  min_ansible_version: 2.4
  platforms:
    - name: ubuntu
      versions:
        - trusty
        - xenial
        - bionic
  galaxy_tags: []
  dependencies:
    - { role: role }
    - { role: username.myrole }
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers molecule/create.yml is correctly formatted 1`] = `
"---
- name: Create
  hosts: localhost
  connection: local
  gather_facts: false
  no_log: \\"{{ not lookup('env', 'MOLECULE_DEBUG') | bool }}\\"
  tasks:
    - name: Create molecule instance(s)
      lxd_container:
        name: \\"{{ item.name }}\\"
        state: started
        source:
          type: image
          mode: pull
          server: https://images.linuxcontainers.org
          protocol: lxd
          alias: \\"{{ item.alias }}/amd64\\"
        profiles: [\\"default\\"]
        wait_for_ipv4_addresses: true
        timeout: 600
      with_items: \\"{{ molecule_yml.platforms }}\\"

    - name: Install Python in container
      delegate_to: \\"{{ item.name }}\\"
      raw: apt-get install -y python
      with_items: \\"{{ molecule_yml.platforms }}\\"
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers molecule/default/molecule.yml is correctly formatted 1`] = `
"---
dependency:
  name: galaxy
driver:
  name: lxd
lint:
  name: yamllint
platforms:
  - name: ubuntu_trusty
    alias: 'ubuntu:14.04'
  - name: ubuntu_xenial
    alias: 'ubuntu:16.04'
  - name: ubuntu_bionic
    alias: 'ubuntu:18.04'

provisioner:
  name: ansible
  playbooks:
    create: ../create.yml
    destroy: ../destroy.yml
  lint:
    name: ansible-lint
scenario:
  name: default
verifier:
  name: testinfra
  lint:
    name: flake8
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers molecule/default/playbook.yml is correctly formatted 1`] = `
"---
- name: Converge
  hosts: all
  roles:
      - name: test-role
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers molecule/default/tests/test_default.py is correctly formatted 1`] = `
"import os

import testinfra.utils.ansible_runner

testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
    os.environ['MOLECULE_INVENTORY_FILE']).get_hosts('all')


def test_example(host):
    file = host.file('/etc/hosts')

    assert file.exists
    assert file.user == 'root'
    assert file.group == 'root'
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers molecule/destroy.yml is correctly formatted 1`] = `
"---
- name: Destroy
  hosts: localhost
  connection: local
  gather_facts: false
  no_log: \\"{{ not lookup('env', 'MOLECULE_DEBUG') | bool }}\\"
  tasks:
    - name: Destroy molecule instance(s)
      lxd_container:
        name: \\"{{ item.name }}\\"
        state: absent
        force_stop: \\"{{ item.force_stop | default(true) }}\\"
      with_items: \\"{{ molecule_yml.platforms }}\\"
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers tasks/main.yml is correctly formatted 1`] = `
"---
"
`;

exports[`generator-molecule-lxd-role:app when all the prompts have answers vars/main.yml is correctly formatted 1`] = `
"---
"
`;

exports[`generator-molecule-lxd-role:app when targetting another distribution than Ubuntu molecule/default/molecule.yml formats the platform list properly 1`] = `
"---
dependency:
  name: galaxy
driver:
  name: lxd
lint:
  name: yamllint
platforms:
  - name: debian_jessie
    alias: 'images:debian/8'
  - name: debian_stretch
    alias: 'images:debian/9'

provisioner:
  name: ansible
  playbooks:
    create: ../create.yml
    destroy: ../destroy.yml
  lint:
    name: ansible-lint
scenario:
  name: default
verifier:
  name: testinfra
  lint:
    name: flake8
"
`;

exports[`generator-molecule-lxd-role:app when there are no dependencies README.md skips the dependencies section 1`] = `
"Test role
=========
[![Build Status](https://travis-ci.org/coaxial/ansible-role-lxd.svg?branch=master)](https://travis-ci.org/coaxial/ansible-role-lxd)

This is a test description for the role.

Requirements
------------

This role makes some assumptions.

Role Variables
--------------

Name | Default | Possible values | Description
---|---|---|---
\`example_variable\` | \`true\` | \`true\` or \`false\` | This is an example to populate the table.


Example Playbook
----------------

\`\`\`yaml
- hosts: all
  roles:
    - test-role
\`\`\`

License
-------

MIT

Author Information
------------------

Test author ([Test organization](https://example.org))"
`;

exports[`generator-molecule-lxd-role:app when there are no dependencies meta/main.yml formats the metadata properly 1`] = `
"---
galaxy_info:
  author: Test author
  description: This is a test description for the role.
  company: Test organization
  license: MIT
  min_ansible_version: 2.4
  platforms:
    - name: ubuntu
      versions:
        - trusty
        - xenial
        - bionic
  galaxy_tags: []
  dependencies: []
"
`;

exports[`generator-molecule-lxd-role:app when there are no requirements README.md skips the requirents section 1`] = `
"Test role
=========
[![Build Status](https://travis-ci.org/coaxial/ansible-role-lxd.svg?branch=master)](https://travis-ci.org/coaxial/ansible-role-lxd)

This is a test description for the role.

Role Variables
--------------

Name | Default | Possible values | Description
---|---|---|---
\`example_variable\` | \`true\` | \`true\` or \`false\` | This is an example to populate the table.

Dependencies
------------

- role
- username.myrole

Example Playbook
----------------

\`\`\`yaml
- hosts: all
  roles:
    - test-role
\`\`\`

License
-------

MIT

Author Information
------------------

Test author ([Test organization](https://example.org))"
`;

exports[`generator-molecule-lxd-role:app when there are no requirements meta/main.yml formats the metadata properly 1`] = `
"---
galaxy_info:
  author: Test author
  description: This is a test description for the role.
  company: Test organization
  license: MIT
  min_ansible_version: 2.4
  platforms:
    - name: ubuntu
      versions:
        - trusty
        - xenial
        - bionic
  galaxy_tags: []
  dependencies:
    - { role: role }
    - { role: username.myrole }
"
`;

exports[`generator-molecule-lxd-role:app when there is no organization README.md formats the author information properly 1`] = `
"Test role
=========
[![Build Status](https://travis-ci.org/coaxial/ansible-role-lxd.svg?branch=master)](https://travis-ci.org/coaxial/ansible-role-lxd)

This is a test description for the role.

Requirements
------------

This role makes some assumptions.

Role Variables
--------------

Name | Default | Possible values | Description
---|---|---|---
\`example_variable\` | \`true\` | \`true\` or \`false\` | This is an example to populate the table.

Dependencies
------------

- role
- username.myrole

Example Playbook
----------------

\`\`\`yaml
- hosts: all
  roles:
    - test-role
\`\`\`

License
-------

MIT

Author Information
------------------

Test author ([https://example.org](https://example.org))
"
`;

exports[`generator-molecule-lxd-role:app when there is no organization meta/main.yml formats the metadata properly 1`] = `
"---
galaxy_info:
  author: Test author
  description: This is a test description for the role.
  license: MIT
  min_ansible_version: 2.4
  platforms:
    - name: ubuntu
      versions:
        - trusty
        - xenial
        - bionic
  galaxy_tags: []
  dependencies:
    - { role: role }
    - { role: username.myrole }
"
`;

exports[`generator-molecule-lxd-role:app when there is no website README.md formats the author information properly 1`] = `
"Test role
=========
[![Build Status](https://travis-ci.org/coaxial/ansible-role-lxd.svg?branch=master)](https://travis-ci.org/coaxial/ansible-role-lxd)

This is a test description for the role.

Requirements
------------

This role makes some assumptions.

Role Variables
--------------

Name | Default | Possible values | Description
---|---|---|---
\`example_variable\` | \`true\` | \`true\` or \`false\` | This is an example to populate the table.

Dependencies
------------

- role
- username.myrole

Example Playbook
----------------

\`\`\`yaml
- hosts: all
  roles:
    - test-role
\`\`\`

License
-------

MIT

Author Information
------------------


Test author (Test organization)
"
`;

exports[`generator-molecule-lxd-role:app when there is no website meta/main.yml formats the metadata properly 1`] = `
"---
galaxy_info:
  author: Test author
  description: This is a test description for the role.
  company: Test organization
  license: MIT
  min_ansible_version: 2.4
  platforms:
    - name: ubuntu
      versions:
        - trusty
        - xenial
        - bionic
  galaxy_tags: []
  dependencies:
    - { role: role }
    - { role: username.myrole }
"
`;
